### ISSUE TITLE
Allow millis less than 4ms

### ISSUE THREAD

**Issue by Shinigami92** [Shinigami92/visualised-sorting-algorithm#1]
@schulefant would like to configure millis with less than 4ms

I need to try if this would have any affect at all

The line that would need to be changed would be here:
https://github.com/Shinigami92/visualised-sorting-algorithm/blob/1cf0002e811d3b23f950d3f550c7350e15e92ee0/src/App.pug#L35

However, the project is in a really old state, and I would need to republish the GitHub page. Because there is no CI/CD process yet, I might also need to set this up.
And then pray that the dependencies do not have problems.

Not sure when I find the time yet, because I have also more than 5 other more high prio projects in my current non-company pipeline.

**Comment by Shinigami92** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/setInterval#delay_restrictions

<img width="882" alt="image" src="https://github.com/Shinigami92/visualised-sorting-algorithm/assets/7195563/c10a708b-9601-41b4-8682-7a5926a55efe">

The official MDN points out that 4 ms is the allowed minimum.
The browser would pin it to 4 ms if it would be set to 1, 2 or 3.
If set to 0 it could either sort it instantly without any animation, or it would also pin it to 4 ms.

TBH, the only thing I could try is to rewrite the project and use https://vueuse.org/core/useRafFn/#useraffn instead of 

https://github.com/Shinigami92/visualised-sorting-algorithm/blob/1cf0002e811d3b23f950d3f550c7350e15e92ee0/src/shared/AbstractSortService.ts#L74

But I'm absolute unsure if this would help.
The other way would be to re-write everything to use a WebGPU shader. But no thx 😆

**Comment by ST-DDT** [comment]
How about bulking changes. The issue isn't the 1 vs 4ms limit. The issue is the `O(n²)` that e.g. bubble sort uses.
To avoid that you could use some kind of bulk operations.

````ts
let i= 0; // persistent state to even off fractional bulk values

while (running) {
  for ( i = i % bulk; i<bulk; i++) {
    progress();
  }
  sleep();
}
````

You could theoretically calculate the bulk count by separating the slider from the actual sleep time.

````ts
if (interval >= 4ms) {
  bulk = 1;
  sleep = interval;
} else {
  bulk = interval/4
  sleep = 4
}
````

In the UI you could make the difference visible by adding a ~ to the value.

But since the interval between 0ms and 4ms is very small I would suggest using a logarithmic scale anyway.
There is hardly any difference between using 9 or 10ms delays, but it makes selecting 10ms way easier.

And while you are at it, you could also add a (hide-able) estimated time value.
This doesn't have to be updated, but you could put some kind of of calculation there `500² * 10ms = ~42 mins` in the worst case (though you could half that to account for the `best+worst/2` estimation)

**Comment by Shinigami92** [comment]
Yeah, sounds like a plan.

**Comment by schulefant** [comment]
I'm really sorry to have kicked off a landslide. I thought it would be really simple. That goes to show - doesn't it? SORRY. But I do know that a problem eats at you, until you have solved it. Sorry again.
But this solution seems manageable. Is it? How much time would it take to implement? I do not want to inconvenience you out of proportion.

**Comment by Shinigami92** [comment]
> I'm really sorry to have kicked off a landslide. I thought it would be really simple. That goes to show - doesn't it? SORRY. But I do know that a problem eats at you, until you have solved it. Sorry again. But this solution seems manageable. Is it? How much time would it take to implement? I do not want to inconvenience you out of proportion.

It's okay, you just had a problem (as customer) and now we need to find a solution. But because this is not paid work, I need to do it in my free-time.
The idea that @ST-DDT provided sound like an interesting enhancement and I can just gain more experience by trying out to implement it.

But also because this project is 4 years old, I would like to fully rewrite it to todays standards.

I think I could find time around eastern vacation, if no other project will take precedence.

**Comment by Shinigami92** [comment]
The full rewrite to newest standard was made in https://github.com/Shinigami92/visualised-sorting-algorithm/commit/a9d231d2bff9e0378bd136f3d8d1e40d1cb94075 in two sessions a 1-2h

I can start from here to attempt the requested suggestion, but I might progress slowly with so much children and family around me 😅

**Comment by schulefant** [comment]
Lieber Christopher,
vielen, vielen Dank - es bubbled wieder. 😊😀😍
Falls ich irgendwann wieder mal in Hamburg sein sollte, werde ich Dich zum Essen einladen.


Beste Grüße,

Alexandra Rollins



________________________________
Von: Shinigami ***@***.***>
Gesendet: Sonntag, 31. März 2024 11:06
An: Shinigami92/visualised-sorting-algorithm ***@***.***>
Cc: Rollins, Alexandra ***@***.***>; Mention ***@***.***>
Betreff: Re: [Shinigami92/visualised-sorting-algorithm] Allow millis less than 4ms (Issue #1)


Closed #1<https://github.com/Shinigami92/visualised-sorting-algorithm/issues/1> as completed via #2<https://github.com/Shinigami92/visualised-sorting-algorithm/pull/2>.
